(DOCS-4353) Add page for creating an agent-integration for marketplace (#16315)
DataDog/documentation: content/en/developers/marketplace/marketplace-agent-integration.md

---
title: Create an Agent-based Integration for Marketplace
type: documentation

> description: Learn how to create an Agent-based integration for the marketplace repo.

Reviewer: ```suggestion
description: Learn how to create an Agent-based integration for the Datadog Marketplace repository.
```

Assuming you've installed [the Agent Integration Developer Tool][3], configure the tool for the `marketplace` repo:

Set `marketplace` as the default working repository:

```
ddev config set marketplace $HOME/dd/marketplace
ddev config set repo marketplace
```

If you used a directory other than `$HOME/dd` to clone the marketplace directory, use the following command to set your working repository:

```
ddev config set marketplace <PATH/TO/MARKETPLACE>
ddev config set repo marketplace
```

## Create your integration

Once you've downloaded Docker, installed an appropriate version of Python, and prepared your development environment, you can get started with creating an Agent-based integration.

### Create scaffolding for your integration

The `ddev create` command runs an interactive tool that creates the basic file and path structure (or "scaffolding") necessary for a new Agent-based integration.

To generate the scaffolding for an Agent-based integration:
1. Make sure you're inside the `marketplace` directory:
   ```
   cd $HOME/dd/marketplace
   ```

2. Run the `ddev` command with the `-t tile` option:
   ```
   ddev create "<INTEGRATION NAME>"
   ```

## Write an Agent Check

> At the core of each Agent-based integration is an *Agent Check* that periodically collects information and sends it to Datadog. Checks inherit their logic from the `AgentCheck` base class and have the the following requirements:

Reviewer: ```suggestion
At the core of each Agent-based integration is an *Agent Check* that periodically collects information and sends it to Datadog. Checks inherit their logic from the `AgentCheck` base class and have the following requirements:
```

> - Integrations running on the Datadog Agent v7 and later must be compatible with Python 3; however, Agents v5 and v6 still use Python 2.7.

Reviewer: ```suggestion
- Integrations running on the Datadog Agent v7 or later must be compatible with Python 3; however, Agents v5 and v6 still use Python 2.7.
```

    instance = {}
    c = AwesomeCheck('awesome', {}, [instance])

    # empty instance
    with pytest.raises(ConfigurationError):
        c.check(instance)

    # only the url
    with pytest.raises(ConfigurationError):
        c.check({'url': 'http://foobar'})

    # only the search string
    with pytest.raises(ConfigurationError):
        c.check({'search_string': 'foo'})

    # this should not fail
    c.check({'url': 'http://foobar', 'search_string': 'foo'})
{{< /code-block >}}

`pytest` has the concept of markers that can be used to group tests into categories. In the example above, `test_config` is marked as a `unit` test.

The scaffolding is set up to run all the tests located in `INTEGRATION_NAME/tests`.

To run the tests, run:

```
ddev test INTEGRATION_NAME
```

### Write an integration test

The [unit test above](#write-a-unit-test) doesn't check the collection logic. To test the logic, you need to create an environment for an integration test and write an integration test.

#### Create an environment for the integration test

To create an environment for your integration test, create a Docker Compose file at `INTEGRATION_NAME/tests/docker-compose.yml`

> For the example check, the toolkit uses `docker` to spin up an Nginx container and allows the check to retrieve the welcome page. The `docker-compose.yml` file has the following contents:

Reviewer: ```suggestion
For the example check, the toolkit uses `docker` to spin up an NGINX container and allows the check to retrieve the welcome page. The `docker-compose.yml` file has the following contents:
```

version: "3"

services:
  nginx:
    image: nginx:stable-alpine
    ports:
      - "8000:80"

{{< /code-block >}}

Next, define the environment for the integration test at `INTEGRATION_NAME/tests/conftest.py`.

Here is an example integration test for the `awesome` check:

{{< code-block lang="python" filename="conftest.py" collapsible="true" >}}
import os

import pytest

from datadog_checks.dev import docker_run, get_docker_hostname, get_here

URL = 'http://{}:8000'.format(get_docker_hostname())
SEARCH_STRING = 'Thank you for using nginx.'
INSTANCE = {'url': URL, 'search_string': SEARCH_STRING}

@pytest.fixture(scope='session')
def dd_environment():
    compose_file = os.path.join(get_here(), 'docker-compose.yml')

    # This does 3 things:
    #
    # 1. Spins up the services defined in the compose file
    # 2. Waits for the url to be available before running the tests
    # 3. Tears down the services when the tests are finished
    with docker_run(compose_file, endpoints=[URL]):
        yield INSTANCE

@pytest.fixture
def instance():
    return INSTANCE.copy()
{{< /code-block >}}

#### Add an integration test

> After you've setup an environment for the integration test, add an integration test to the `INTEGRATION_NAME/tests/test_INTEGRATION_NAME.py` file:

Reviewer: ```suggestion
After you've set up an environment for the integration test, add an integration test to the `INTEGRATION_NAME/tests/test_INTEGRATION_NAME.py` file:
```

Logos
: Add at least one SVG file which can be used in light and dark modes in a `logos` folder nested in the `assets` directory, or add the file(s) directly to the `assets` directory.<br><br>Technology Partners are responsible for the licensing of submitted logos.

CHANGELOG
: Document release notes and version information in the `CHANGELOG.md` file using the following format: `1.0.0 / YYYY-MM-DD`. This information is displayed in the **Release Notes** tab of the integration tile.<br><br>Technology Partners can add releases and version updates in descending order (latest version at the top).

CODEOWNERS
: The `CODEOWNERS` file belongs in the shared `.github` directory and defines the individuals or teams responsible for maintaining the content and source code in the Marketplace repository.<br><br>For information on CODEOWNER syntax, see the [GitHub code owners documentation][19].

End User License Agreement (EULA)
: Add the `eula.pdf` file in the `assets` directory.<br><br>Technology Partners are responsible for adding the EULA.

### README

{{% integrations_readme %}}

### Media carousel

{{% integrations_media_carousel %}}

## Build the wheel

> The `pyproject.toml` file provides the metadata that is used to package and build the wheel. The wheel contains the files necessary for the functioning of the integration itself, which includes the Check, configuration example file, and artifacts generated during the build of the wheel.

Reviewer: "Functioning of the integration" sounded off to me, but feel free to reword or ignore!

```suggestion
The `pyproject.toml` file provides the metadata that is used to package and build the wheel. The wheel contains the files necessary for the integration to function itself, and includes the Check, configuration example file, and artifacts generated during the build of the wheel.
```

All additional elements, including the metadata files, are not meant to be contained within the wheel, and are used elsewhere by the Datadog platform and ecosystem. To learn more about Python packaging, see [Packaging Python Projects][20].

Once your `pyproject.toml` is ready, create a wheel:

- (Recommended) With the `ddev` tooling: `ddev release build <INTEGRATION_NAME>`.
- Without the `ddev` tooling: `cd <INTEGRATION_DIR> && pip wheel . --no-deps --wheel-dir dist`.

## Install the wheel

> The wheel is installed using the Agent `integration` command, available in [Agent v6.10.0 and up][21]. Depending on your environment, you may need to execute this command as a specific user or with specific privileges:

Reviewer: ```suggestion
The wheel is installed using the Agent `integration` command, available in [Agent v6.10.0 or later][21]. Depending on your environment, you may need to execute this command as a specific user or with specific privileges:
```

  <pre><code>& "C:\Program Files\Datadog\Datadog Agent\embedded\agent.exe" integration install -w /path/to/wheel.whl</code></pre>

</details>

<details open>
  <summary>Agent<code>v6.12</code> or later</summary>

  <pre><code>& "C:\Program Files\Datadog\Datadog Agent\bin\agent.exe" integration install -w /path/to/wheel.whl</code></pre>
</details>

{{% /tab %}}
{{< /tabs >}}

## Open a pull request

Before you open a pull request, run the following command to catch any problems with your integration:

```
ddev validate all <INTEGRATION_NAME>
```

Push up your feature branch and open a pull request that contains your integration tile's asset files (including images) in the [`marketplace` repository][7]. The Marketplace repository does not allow forks. For instructions on creating a clone of the repo, see the [Set up section](#set-up-a-directory-and-clone-the-marketplace-repository). After you've created your pull request, automatic checks run in Azure DevOps pipelines to verify that your pull request is in good shape and contains all the required content to be updated.

To request access to the Azure DevOps pipeline, leave a comment in the pull request requesting access.

> After you've created your pull request, automatic checks run to verify that your pull request is in good shape and contains all the required content to be updated.

Reviewer: Extra sentence here I think

```suggestion
```

The wheel is installed using the Agent `integration` command, available in [Agent v6.10.0 and up][21]. Depending on your environment, you may need to execute this command as a specific user or with specific privileges:

{{< tabs >}}
{{% tab "Linux (as dd-agent)" %}}
<pre><code>sudo -u dd-agent datadog-agent integration install -w /path/to/wheel.whl</code></pre>
{{% /tab %}}

{{% tab "Mac OS (as admin)" %}}
<pre><code>sudo datadog-agent integration install -w /path/to/wheel.whl</code></pre>
{{% /tab %}}

{{% tab "Windows PowerShell" %}}

Ensure that your shell session has _administrator_ privileges.

<details>
  <summary>Agent <code>v6.11</code> or earlier</summary>

  <pre><code>& "C:\Program Files\Datadog\Datadog Agent\embedded\agent.exe" integration install -w /path/to/wheel.whl</code></pre>

</details>

<details open>
  <summary>Agent<code>v6.12</code> or later</summary>

  <pre><code>& "C:\Program Files\Datadog\Datadog Agent\bin\agent.exe" integration install -w /path/to/wheel.whl</code></pre>
</details>

{{% /tab %}}
{{< /tabs >}}

## Open a pull request

Before you open a pull request, run the following command to catch any problems with your integration:

```

> ddev validate all changes

Author: ```suggestion
ddev validate all <INTEGRATION_NAME>
```
